### PR TITLE
WA-4090 Cashflow reports API changes

### PR DESCRIPTION
**Pull request by VindhyaRajagopal**
## The problem ##

[WA-4090](https://frollo.atlassian.net/browse/WA-4090)
[ Doc Reference](https://frollo.atlassian.net/wiki/spaces/DEV/pages/3030908999/Cashflow+Reports+API+Money+In+Money+Out+-+Credit+Debit)

## Checklist ##

Use the :white_check_mark: emoji for when it's been done; :no_entry: when it still needs work;
strikethrough when an item of the checklist does not apply to the changes.

* Is the destination branch correct & intended?:white_check_mark:
* Is the code synced to latest (if applicable)?:white_check_mark:
* Has the SDK example app build passed locally?:white_check_mark:
* Is the correct JIRA ticket referenced?:white_check_mark:
* Is the correct API docs/Confluence Page referenced?
* Have all the AC's in the ticket passed?:white_check_mark:
* Have you followed proper naming conventions in the code?:white_check_mark:
* Has the code been properly covered with unit & android tests?
* Have the full test suite (unit & android tests) passed locally?
* Has the changelog.md, packages.md, overview.md, extras.md been updated?:white_check_mark:
* Has the dokka task passed?NA
* Have you committed database schema changes (if applicable)?NA

### CODE REVIEW

File: src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt [hunk 1]
```diff
@@ -293,6 +296,38 @@ class Reports(network: NetworkService, internal val db: SDKDatabase, private val
         }
     }
 
+    /**
+     * Fetch tag reports from the host
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt:300]: cashflow reports

File: src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt [hunk 2]
```diff
@@ -293,6 +296,38 @@ class Reports(network: NetworkService, internal val db: SDKDatabase, private val
...
+     *
+     * @param fromDate Start date in the format yyyy-MM-dd to fetch reports from (inclusive). See [Report.DATE_FORMAT_PATTERN]
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt:302]: CashflowReport. DATE_FORMAT_PATTERN

File: src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt [hunk 3]
```diff
@@ -293,6 +296,38 @@ class Reports(network: NetworkService, internal val db: SDKDatabase, private val
...
+     * @param toDate End date in the format yyyy-MM-dd to fetch reports up to (inclusive). See [Report.DATE_FORMAT_PATTERN]
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt:303]: CashflowReport. DATE_FORMAT_PATTERN

File: src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt [hunk 4]
```diff
@@ -293,6 +296,38 @@ class Reports(network: NetworkService, internal val db: SDKDatabase, private val
...
+     * @param period Period that reports should be broken down by
+     * @param completion Optional completion handler with optional error if the request fails
+     */
+    fun fetchCashflowReports(
+        fromDate: String,
+        toDate: String,
+        period: TransactionReportPeriod,
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt:310]: are these parameters optional or all are required?

File: src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt [hunk 5]
```diff
@@ -293,6 +296,38 @@ class Reports(network: NetworkService, internal val db: SDKDatabase, private val
...
+            toDate = toDate
+        ).enqueue { resource ->
+
+            if (resource.status == Resource.Status.ERROR) {
+                Log.e("$TAG#fetchCashflowReports", resource.error?.message)
+                completion?.invoke(Resource.error(resource.error))
+            }
+            completion?.invoke(
+                resource.map { response ->
+                    response?.toCashflowReports()
```
Comment (deepakmn86) [inline src/main/kotlin/us/frollo/frollosdk/reports/Reports.kt:325]: you can just do `response.data`

File: src/sharedTest/kotlin/us/frollo/frollosdk/model/TestCashflowReportData.kt [hunk 6]
```diff
@@ -0,0 +1,3 @@
+package us.frollo.frollosdk.model
+
+class TestCashflowReportData
```
Comment (deepakmn86) [inline src/sharedTest/kotlin/us/frollo/frollosdk/model/TestCashflowReportData.kt:3]: do we need this?

File: src/test/kotlin/us/frollo/frollosdk/mapping/ReportsMappingTest.kt [hunk 7]
```diff
@@ -76,4 +77,15 @@ class ReportsMappingTest {
         assertEquals("living", model.name)
         assertEquals(2, model.transactionIds?.size)
     }
+
+    @Test
+    fun testCashflowReportResponseToCashflowReport() {
```
Comment (deepakmn86) [inline src/test/kotlin/us/frollo/frollosdk/mapping/ReportsMappingTest.kt:82]: not needed if we remove the mapping function itself

File: src/sharedTest/kotlin/us/frollo/frollosdk/model/TestDataReport.kt [hunk 8]
```diff
@@ -89,3 +91,17 @@ internal fun testReportAccountBalanceData(
 
     return report
 }
+
+internal fun testCashflowReportResponseData(): CashflowReportResponse {
```
Comment (deepakmn86) [inline src/sharedTest/kotlin/us/frollo/frollosdk/model/TestDataReport.kt:95]: not needed if we remove the mapping function itself?

**Review by deepakmn86 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/us/frollo/frollosdk/model/coredata/reports/CashflowReport.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    @SerializedName("date")val date: String,
+
+    /** Sum of credit transactions */
+    @SerializedName("credits")val credits: String,
+
+    /** Sum of debit transactions */
+    @SerializedName("debits")val debits: String,
+
+    /** Balance of credit and debit transactions summed together */
+    @SerializedName("balance")val balance: String
```
[the remark this answers is not in the document]
Comment (VindhyaRajagopal) [reply src/main/kotlin/us/frollo/frollosdk/model/coredata/reports/CashflowReport.kt:35]: All these are required as per doc.
Screen Shot 2022-11-14 at 12 05 40 pm

**Review by deepakmn86 — APPROVED** [review]

**Pull request merged**